Added script for auto updating the pricing with the pricings of openrouter.ai

vietpham1911 opened the pull request:
## Added script for auto updating the pricing with the pricings of openrouter.ai

<!-- e.g. "Implement user authentication feature" -->

## Relevant issues
#2407

## Type

<!-- Select the type of Pull Request -->
<!-- Keep only the necessary ones -->

🆕 New Feature

## Changes
I added a script for automatically updating the pricings of the llm models with the prices of openrouter.ai. When should the script get triggered?

<!-- Test procedure -->

vietpham1911 committed:
Added script for auto updating the pricing with the pricings of openrouter.ai

Merlinvt commented:
The script does not add new models right? It only updates existing ones ? It would be really cool if new models would be added automatically. One of the great things about open router is, that they have  all the new models super fast. Its amazing to try out SoTA models, by just changing the model name.

Merlinvt commented on auto_update_pricing.py:
```diff
@@ -0,0 +1,93 @@
+        print("Error fetching data from URL:", e)
+        return None
+
+def update_local_data(local_data, remote_data):
```
I modified the function to update the models with additional fields and include new models. However, this update seems to have unintentionally changed all the pricing information to scientific notation. I don't know how to fix that.
def update_local_data(local_data, remote_data):
    """
    Updates local data with information fetched remotely.
    Args:
        local_data (dict): Local data to be updated.
        remote_data (dict): Remote data fetched from an API.
    """
    # Prepare set of existing openrouter model names for strict checking
    existing_models = {model_name for model_name in local_data if model_name.startswith("openrouter/")}

    for model in remote_data["data"]:
        model_suffix = model["id"]
        model_name = f"openrouter/{model_suffix}"

        # Calculate per token and per image costs
        input_cost_per_token = float(model["pricing"]["prompt"]) 
        output_cost_per_token = float(model["pricing"]["completion"]) 
        input_cost_per_image = float(model["pricing"]["image"]) if "image" in model["pricing"] else None


        # Preparing model data dictionary
        model_data = {
            "max_tokens": model["context_length"],
            "input_cost_per_token": float(input_cost_per_token),
            "output_cost_per_token": float(output_cost_per_token),
            "litellm_provider": "openrouter",
            "mode": "chat",
        }

        if model["per_request_limits"] is not None:
            if "prompt_tokens" in model["per_request_limits"]:
                model_data["max_input_tokens"] = int(model["per_request_limits"]["prompt_tokens"])
            if "completion_tokens" in model["per_request_limits"]:
                model_data["max_output_tokens"] = int(model["per_request_limits"]["completion_tokens"])

        if input_cost_per_image is not None and input_cost_per_image != 0.0:
            model_data["input_cost_per_image"] = float(input_cost_per_image)

        if model["architecture"]["modality"] == "multimodal":
            model_data["supports_vision"] = True

        if model_name in existing_models:
            # Update existing model with the "openrouter/" prefix
            local_data[model_name].update(model_data)
        else:
            # Add new model with the "openrouter/" prefix
            local_data[model_name] = model_data

vietpham1911 replied:
Adding new models is a good idea. I will see if I can fix this :) Grüße aus Australien! (Ich mache gerade ein Auslandssemester und komme aus Stuttgart hahah)

Merlinvt reviewed (changes requested):
Looking forward to this pull request being merged :) thanks Viet! Grüße aus Kassel !

ishaan-jaff commented:
It should get triggered in our ci/cd pipeline - under config.yml

vietpham1911 committed:
Add the models of openrouter that are not in the json file with the pricings of the models yet

vietpham1911 committed:
Bug fix: The prior code removed values like "supports_function_calling" or "supports_vision". This is fixed now.
Nothing is being removed. Only the pricings and the maximum tokens are being edited and new models in openrouter are being added.

vietpham1911 commented:
What if openrouter.ai has models that litellm does not support yet? Should I still add the missing models to the json file?

Merlinvt commented:
Cool, thanks for the change! :)

could you add:
"input_cost_per_image" (model["pricing"]["image"])
and
"supports_vision" (The model does support it if ["architecture"]["modality"] == "multimodal")

I am mostly using the multimodal models through open router, so this would be great if the fields are set.

Merlinvt commented:
> What if openrouter.ai has models that litellm does not support yet? Should I still add the missing models to the json file?

Should be fine right? I don't think l there is any implementation that is needed for a model. The usage of open router is the same regardless of the model, you just change the model parameter. 
I might be wrong of course.

Merlinvt commented:
would you mind just adding the input_cost_per_image and supports_vision ? I think then we can ping someone to have a look at it. I really want to use some of the new models in open routerthrough litellm :)  Getting a little impatient, sorry for that.
I will check who merged by pull request about the model updating and ping him.

krrishdholakia commented:
Hey @vietpham1911 @Merlinvt let me know when this is ready for review! 

Excited to merge this is in! 

Re: frequency of updates - can we do a daily check?

vietpham1911 commented:
> would you mind just adding the input_cost_per_image and supports_vision ? I think then we can ping someone to have a look at it. I really want to use some of the new models in open routerthrough litellm :) Getting a little impatient, sorry for that. I will check who merged by pull request about the model updating and ping him.

I have urgent assignemnts right now, but I think I can keep working on it next week again.

Merlinvt commented:
Thanks so much :) looking forward to the new models on openrouter. In particular Gemini 1.5 flash and GPT4o.

paneru-rajan commented:
> It should get triggered in our ci/cd pipeline - under config.yml

@ishaan-jaff / @krrishdholakia 

Here're few things we need to have this wrapped up:

**Question 1:**
Should we place this file under `ci_cd` folder?

**Question 2:**
I think we should be using `aiohttp` instead of requests as `requests` is not available in the pipeline

**Question 3:**
Does the placement of the code looks good to you in the `.circleci/config.yml` file?
<img width="1038" alt="image" src="https://github.com/BerriAI/litellm/assets/4735661/22481789-4c88-4e45-b405-a5f2d03bb79f">

CC: @vietpham1911

paneru-rajan committed:
updated the config to run the script

vietpham1911 commented:
The float values are being converted into scientific notation `1.2e-07` in the json file. Is it okay to use scientific notation? Or do you prefer to have the values as a string "0.00000012"?

vietpham1911 committed:
Wrote the code more clean and more efficient. Furthermore, the code adds the input_cost_per_image and supports_vision to the models if the two values exist.

vietpham1911 committed:
Implemented feature to add input_cost_per_image

paneru-rajan committed:
resolve conflict

Merlinvt commented:
@krrishdholakia 
Looks good for me it has the multimodal flag and pricing :)
Thanks @vietpham1911

krrishdholakia commented:
Great work @vietpham1911 @Merlinvt @paneru-rajan 

Can someone walk me through, 

- how this needs to be setup to check for openrouter pricing? 
- how are changes submitted? Is this via a pr?

paneru-rajan commented:
> Great work @vietpham1911 @Merlinvt @paneru-rajan
> 
> Can someone walk me through,
> 
> * how this needs to be setup to check for openrouter pricing?
> * how are changes submitted? Is this via a pr?

Based on the suggestion of @ishaan-jaff it will run during the ci/cd pipeline. This will not commit the changes in the repo, it will update the file at the time of docker build.

<img width="925" alt="image" src="https://github.com/BerriAI/litellm/assets/4735661/1f6fd8ae-008d-4893-8b73-3f7fed8f69f3">

@krrishdholakia / @ishaan-jaff

Merlinvt commented:
What if you use Litellm without docker ?

paneru-rajan committed:
update price & context window file daily, commit and  push to the main

paneru-rajan committed:
Daily JSON update: 2024-05-19

paneru-rajan commented:
**Let me know what we should do:**

I think a github workflow to do this will be ideal than circleci, which will  run daily/week to:
1. create a PR (which then need to be merged manually)
2. Merge in main (not recommended approach)

I can write a near working github workflow but will not be able to test, @krrishdholakia / @ishaan-jaff you need to take it from there.

CC: @Merlinvt

paneru-rajan committed:
reverted the changes made in circleci

paneru-rajan committed:
created github worflow to update the file and create a pull-request

paneru-rajan commented:
@ishaan-jaff / @krrishdholakia I've added a github workflow to update the file and create a pr weekly, I will not be able to test, if everything looks good, can you test it and let me know if there's any changes needed.

@Merlinvt this will resolve the concern of yours, this will run on schedule rather than at ci/cd.

paneru-rajan committed:
moved script to the workflow folder

Merlinvt commented:
Awesome thanks ;)

Merlinvt commented:
I am camping right now, so i can test it on a computer. When can this be merged ?

Merlinvt commented:
@vietpham1911 looks great. I used the branch version for my use case, and it seems to be working.

However, there are two things that would be amazing if you could update them:

1.If the model is not multimodal, there is no info needed for input_cost_per_image. Right now it's 0.0 for a lot of models.
2. Could you add the field max_output_tokens (is it max_completion_tokens in OpenRouter?).

I just realized that the project I am using requires max_output_tokens, so it would be great if you could add it.

ishaan-jaff commented on model_prices_and_context_window.json:
```diff
@@ -1,10 +1,10 @@
-        "max_output_tokens": 4096, 
-        "input_cost_per_token": 0.00003,
-        "output_cost_per_token": 0.00006,
+        "max_output_tokens": 4096,
```
please do not edit OpenAI pricing in this PR or with your script. It should only be limited to Open router

paneru-rajan replied:
Hi @ishaan-jaff , I double checked: we are only updating the values of open-router.

Following are **same values**, one in `float` another in `scientific notation`.
<img width="1300" alt="image" src="https://github.com/BerriAI/litellm/assets/4735661/6441d145-0ae1-47bc-be1b-3ea9eec84b23">

This is a default behavior of the `json` package, that converts the float to "scientific notation".

If we want the floating point not "scientific notation" then, the **floating point values will be string type** similar to what OpenRouter is doing `"0.000001125"`:
<img width="663" alt="image" src="https://github.com/BerriAI/litellm/assets/4735661/5307c504-5ebc-4d7a-83d1-8283c102f2bb">

> Because of floating point precision `json` package uses scientific notation.

ishaan-jaff reviewed (changes requested):
great work so far! Please do not edit OpenAI pricing in this PR or with your script. It should only be limited to Open router

paneru-rajan committed:
resolve-conclit: pulled the changes from main

paneru-rajan committed:
ran the script after resolving conflict

paneru-rajan commented:
**Fixed the conflict:** Since constant change are being made to the json file, i had to resolve conflict now and then to this pr.

vietpham1911 committed:
Added the requested features of Merlinvt ("1.If the model is not multimodal, there is no info needed for input_cost_per_image. Right now it's 0.0 for a lot of models.
2. Could you add the field max_output_tokens (is it max_completion_tokens in OpenRouter?).")

vietpham1911 commented:
> @vietpham1911 looks great. I used the branch version for my use case, and it seems to be working.
> 
> However, there are two things that would be amazing if you could update them:
> 
> 1.If the model is not multimodal, there is no info needed for input_cost_per_image. Right now it's 0.0 for a lot of models. 2. Could you add the field max_output_tokens (is it max_completion_tokens in OpenRouter?).
> 
> I just realized that the project I am using requires max_output_tokens, so it would be great if you could add it.

I implemented your two requests :)

krrishdholakia commented:
Hi team - can you please remove your model cost json modifications - model_prices_and_context_window.json

This changes our 'gpt-4' pricing for openai. 

After that, it should be good to merge.

paneru-rajan commented:
> Hi team - can you please remove your model cost json modifications - model_prices_and_context_window.json
> 
> This changes our 'gpt-4' pricing for openai.
> 
> After that, it should be good to merge.

@krrishdholakia like I mentioned above, they are the same value but are in scientific notation. I will make the data type as string then.

https://github.com/BerriAI/litellm/pull/3630#discussion_r1609175083

krrishdholakia commented:
@paneru-rajan no please remove the file. We can accept any changes the script catches via a subsequent PR.

paneru-rajan committed:
reverted with the main to keep them same, and subsequent pr will make these changes

paneru-rajan commented:
> @paneru-rajan no please remove the file. We can accept any changes the script catches via a subsequent PR. The json is used by the package in prod, and we don't want to ship any breaking changes.

Thank you for clarification, I've removed the changes as you've requested.
